### PR TITLE
fix(api): slippage causing negative balances

### PR DESCRIPTION
**Pull request by dtbuchholz**
closes [APP-344](https://linear.app/recall-labs/issue/APP-344/bug-negative-prices-in-trades-table). here's where the `calculateSlippage` method is used, causing the `effectiveFromValueUSD` (used in `toAmount`) and resulting price in the trades table to be negative:
```ts
private calculateTradeAmounts(
    fromAmount: number,
    fromValueUSD: number,
    toPrice: number,
    fromToken: string,
    toToken: string,
  ): { toAmount: number; exchangeRate: number } {
    let toAmount: number;
    let exchangeRate: number;
    let effectiveFromValueUSD: number;

    if (toPrice === 0) {
      // ...
    } else {
      // Normal trade with slippage
      const { effectiveFromValueUSD: calculatedEffectiveValue } =
        calculateSlippage(fromValueUSD);
      effectiveFromValueUSD = calculatedEffectiveValue;
      toAmount = effectiveFromValueUSD / toPrice;
      exchangeRate = toAmount / fromAmount;
      serviceLogger.debug(`[TradeSimulator] Trade calculation details:
          From Token (${fromToken}):
          - Amount: ${fromAmount}
          - USD Value: $${fromValueUSD.toFixed(6)}

          To Token (${toToken}):
          - Price: $${toPrice}
          - Calculated Amount: ${toAmount.toFixed(6)}

          Exchange Rate: 1 ${fromToken} = ${exchangeRate.toFixed(6)} ${toToken}
      `);
    }

    return { toAmount, exchangeRate };
  }
```

### CODE REVIEW

File: apps/api/src/lib/trade-utils.ts [hunk 1]
```diff
@@ -16,13 +16,31 @@ export interface SlippageResult {
  * @returns SlippageResult containing slippage calculations
  */
 export function calculateSlippage(fromValueUSD: number): SlippageResult {
-  // Apply slippage based on trade size
-  const baseSlippage = (fromValueUSD / 10000) * 0.05; // 0.05% per $10,000 (10x lower than before)
-  const actualSlippage = baseSlippage * (0.9 + Math.random() * 0.2); // ±10% randomness (reduced from ±20%)
+  // Apply slippage based on trade size with logarithmic scaling
```
Comment (dtbuchholz) [inline apps/api/src/lib/trade-utils.ts:37]: not sure what the ideal slippage formula is. we'll also throw an error if `effectiveFromValueUSD` is negative

**Review by stbrody — COMMENTED** [review]
tbh I'm not great at reasoning about this kind of stuff.  This feels like something that would benefit from @carsonfarmer taking a look at

File: apps/api/src/lib/trade-utils.ts [hunk 2]
```diff
@@ -16,13 +16,31 @@ export interface SlippageResult {
...
+  // This prevents slippage from exceeding reasonable limits for large trades
+  const MAX_SLIPPAGE = 0.15; // 15% maximum slippage
+
+  // Logarithmic scaling: slippage increases with trade size but at a decreasing rate
+  // Base formula: slippage = k * log(1 + value/scale)
+  // where k = 0.02 (2% base) and scale = 10000 ($10k reference)
+  const baseSlippage = Math.min(
+    0.02 * Math.log10(1 + fromValueUSD / 10000),
```
Comment (carsonfarmer) [inline apps/api/src/lib/trade-utils.ts]: nit: We should probably use `log1p` (with base conversion) for clarity here (i.e., making it clear the intent is log(1+x)) and better behavior for tiny trades. Something like `Math.log1p(fromValueUSD / 10000) / Math.LN10`

File: apps/api/src/lib/trade-utils.ts [hunk 3]
```diff
@@ -16,13 +16,31 @@ export interface SlippageResult {
  * @returns SlippageResult containing slippage calculations
  */
 export function calculateSlippage(fromValueUSD: number): SlippageResult {
-  // Apply slippage based on trade size
-  const baseSlippage = (fromValueUSD / 10000) * 0.05; // 0.05% per $10,000 (10x lower than before)
-  const actualSlippage = baseSlippage * (0.9 + Math.random() * 0.2); // ±10% randomness (reduced from ±20%)
+  // Apply slippage based on trade size with logarithmic scaling
+  // This prevents slippage from exceeding reasonable limits for large trades
```
Comment (carsonfarmer) [inline apps/api/src/lib/trade-utils.ts:38]:
```suggestion
  if (!Number.isFinite(fromValueUSD) || fromValueUSD < 0) {
    throw new Error("fromValueUSD must be a finite, non-negative number");
  }

  // This prevents slippage from exceeding reasonable limits for large trades
```

File: apps/api/src/lib/trade-utils.ts [hunk 4]
```diff
@@ -16,13 +16,31 @@ export interface SlippageResult {
...
+    0.02 * Math.log10(1 + fromValueUSD / 10000),
+    MAX_SLIPPAGE,
+  );
+
+  // Apply randomness: ±10% variation
+  const randomMultiplier = 0.9 + Math.random() * 0.2;
+  const actualSlippage = Math.min(
+    baseSlippage * randomMultiplier,
+    MAX_SLIPPAGE,
+  );
```
Comment (carsonfarmer) [inline apps/api/src/lib/trade-utils.ts:60]: Check me on this, but I think this is a bit clearer and you'll avoid the possibility of a negative.
```suggestion
  const actualSlippage = Math.max(
    0, Math.min(baseSlippage * randomMultiplier, MAX_SLIPPAGE)
  );
```

File: apps/api/src/lib/trade-utils.ts [hunk 5]
```diff
@@ -16,13 +16,31 @@ export interface SlippageResult {
...
   const slippagePercentage = actualSlippage * 100;
 
   // Calculate final amount with slippage
   const effectiveFromValueUSD = fromValueUSD * (1 - actualSlippage);
+  if (effectiveFromValueUSD < 0) {
+    throw new Error("effectiveFromValueUSD is negative");
+  }
```
Comment (carsonfarmer) [inline apps/api/src/lib/trade-utils.ts]: I think if you do the clamping above as I suggested, you no longer need this?

```suggestion
```

File: apps/api/src/lib/__tests__/trade-utils.test.ts [hunk 6]
```diff
@@ -0,0 +1,117 @@
+import { describe, expect, it } from "vitest";
+
+import { calculateSlippage } from "../trade-utils.js";
+
+describe("trade-utils", () => {
```
Comment (carsonfarmer) [inline apps/api/src/lib/__tests__/trade-utils.test.ts:5]: This all looks reasonable and good. Also, yes tests!

Do we need to add tests for negative and non-integer input values? Or, maybe extremely large values (e.g., Number.MAX_SAFE_INTEGER)?
